src/shims/windows/sync.rs

```diff
@@ -135,4 +103,112 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
 
         Ok(())
     }
+
+    fn InitOnceBeginInitialize(
+        &mut self,
+        init_once_op: &OpTy<'tcx, Provenance>,
+        flags_op: &OpTy<'tcx, Provenance>,
+        pending_op: &OpTy<'tcx, Provenance>,
+        context_op: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, Scalar<Provenance>> {
+        let this = self.eval_context_mut();
+        let active_thread = this.get_active_thread();
+
+        let id = this.init_once_get_or_create_id(init_once_op, 0)?;
+        let flags = this.read_scalar(flags_op)?.to_u32()?;
+        let pending_place = this.deref_operand(pending_op)?.into();
+        let context = this.read_pointer(context_op)?;
+
+        if flags != 0 {
+            throw_unsup_format!("unsupported `dwFlags` {flags} in `InitOnceBeginInitialize`");
+        }
+
+        if !this.ptr_is_null(context)? {
+            throw_unsup_format!("non-null `lpContext` in `InitOnceBeginInitialize`");
+        }
+
+        struct Callback<'tcx> {
+            init_once_id: InitOnceId,
+            pending_place: PlaceTy<'tcx, Provenance>,
+        }
+
+        impl<'tcx> VisitTags for Callback<'tcx> {
+            fn visit_tags(&self, visit: &mut dyn FnMut(SbTag)) {
+                let Callback { init_once_id: _, pending_place } = self;
+                pending_place.visit_tags(visit);
+            }
+        }
+
+        impl<'mir, 'tcx> MachineCallback<'mir, 'tcx> for Callback<'tcx> {
+            fn call(&self, this: &mut MiriInterpCx<'mir, 'tcx>) -> InterpResult<'tcx> {
+                let pending = match this.init_once_status(self.init_once_id) {
+                    InitOnceStatus::Uninitialized =>
+                        unreachable!("status should have either been set to begun or complete"),
+                    InitOnceStatus::Begun => this.eval_windows("c", "TRUE")?,
+                    InitOnceStatus::Complete => this.eval_windows("c", "FALSE")?,
+                };
+
+                this.write_scalar(pending, &self.pending_place)?;
+
+                Ok(())
+            }
+        }
+
+        match this.init_once_status(id) {
+            InitOnceStatus::Uninitialized => {
+                this.init_once_begin(id);
+                this.write_scalar(this.eval_windows("c", "TRUE")?, &pending_place)?;
+            }
+            InitOnceStatus::Begun =>
+                this.init_once_enqueue_and_block(
+                    id,
+                    active_thread,
+                    Box::new(Callback { init_once_id: id, pending_place }),
+                ),
```

Comment:
Please move the callback into this match arm. It is confusing to read all that callback code and then realize it is only sometimes relevant.